docs/src/debugging.md

```diff
@@ -72,14 +75,127 @@ ClimaCore.DebugOnly.call_post_op_callback() = false # hide
+triggered the `NaN`. A simple way to find that is to extend this example with
+`Infiltrator`.
+
+#### Infiltrating and Exfiltrating
 
 [Infiltrator.jl](https://github.com/JuliaDebug/Infiltrator.jl) is a simple
 debugging tool for Julia packages.
 
 Here is an example, where we can use Infiltrator.jl to find where NaNs is coming
 from interactively.
 
+##### Infiltrating
+
+Suppose you have a `NaN` in your simulation and what to look at the variables
+right before the expression caused the `NaN`. One of the challenges in this is
+that you probably have a tower of function being called, many of which belong to
+other packages.
+
+Let us simulate this case
+```julia
+import ClimaCore
+using ClimaCore.CommonSpaces
+space = CubedSphereSpace(; radius = 10, n_quad_points = 4, h_elem = 10)
+myrho = ones(space)
+myP = ones(space)
+myu = ones(space)
+
+kinetic_energy(field) = field .* field ./ 2
+other_energy(field) = field ./ sum(field)
+offset_by_one(field) = field .- 1
+
+function specific_energy(rho, P, u)
+    density_without_restmass = offset_by_one(rho)
+    return (kinetic_energy(u) .+ other_energy(P)) ./ density_without_restmass
+end
+
+function renormalized_energy(rho, P, u)
+    energy = specific_energy(rho, P, u)
+    return energy ./ sum(energy)
+end
+
+any(isnan, renormalized_energy(myrho, myP, myu)) # true
+```
+
+To debug this, we first need to identify where the first `NaN` is produced. We
+use `DebugOnly.call_post_op_callback` and infiltrate.
+```julia
+import Infiltrator # must be in your default environment
+ClimaCore.DebugOnly.call_post_op_callback() = true
+function ClimaCore.DebugOnly.post_op_callback(result, args...; kwargs...)
+    has_nans = result isa Number ? isnan(result) : any(isnan, parent(result))
+    has_inf = result isa Number ? isinf(result) : any(isinf, parent(result))
+    @infiltrate has_nans || has_inf
+end
+```
+
+"Infiltrating" means being dropped into a new REPL where in the scope of the
+`@infiltrate` macro. `@infiltrate condition` means that we want to infiltrate
+only when the `condition` is true (in this case `has_nans || has_inf`).
+
+Now, when we run our example, we will see
+```julia
+julia> renormalized_energy(myrho, myP, myu)
+Infiltrating post_op_callback(::ClimaCore.DataLayouts.IJFH{Float64, 4, Array{Float64, 4}}, ::ClimaCore.DataLayouts.IJFH{Float64, 4, Array{Float64, 4}}, ::Vararg{Any}; kwargs::@Kwargs{})
+  at REPL[40]:4
+infil>
+```
+Here, we are dropped into a new REPL with full access to the variables in the scope where the `NaN` occurred. However, because of how `post_op_callback`, this is at a low level within `ClimaCore`, which is typically not useful. Hence, the next step is to type `@trace`, which prints out
+```julia
+[1] post_op_callback(::ClimaCore.DataLayouts.IJFH{…}, ::ClimaCore.DataLayouts.IJFH{…}, ::Vararg{…}; kwargs::@Kwargs{})
+    at REPL[40]:4
+[2] post_op_callback
+    at REPL[40]:1
+[3] copyto!
+    at /home/sbozzolo/repos/ClimaCore.jl/src/DataLayouts/copyto.jl:18
```

Comment:
I tend to try to avoid leaving my personal footprint in code / documentation. Can / should we curate this directory/file?